National Diet Library Online.js

```diff
@@ -0,0 +1,811 @@
+			if (titleNode) {
+				var prefixNote = "";
+				var hasDigital = ZU.xpath(rowNode, ".//p[contains(@class, 'digital')]/a")[0];
+				if (hasDigital) {
+					prefixNote = "[PDF] ";
+				}
+				var childUrl = titleNode.getAttribute("href");
+				var val = titleNode.textContent.trim();
+				info.multiple.urlMap[childUrl] = prefixNote + val;
+			}
+		}
+	} else {
+		// Exactly one anchor with class optAct-linker occurs in item pages. This is the sole
+		// source of the ID needed to build the JSON urls, apart from the URL (which isn't correct
+		// in this phase).)
+		var singleNode = ZU.xpath(doc, "//a[contains(@class, 'optAct-linker')]")[0];
+		if (singleNode) {
+			info.single = {};
+			var key = singleNode.getAttribute('href');
+			key = key.replace(/^.*issToken=/, "").replace(/\&.*$/, "");
+			var url = "detail/" + key;
+			info.single.url = url;
+		} else if (url.match(/detail\//)) {
+			// If we still don't know the page type, check to see if
+			// it's obviously a single by the url.
+			info.single = {};
+			info.single.url = url;
+		}
+	}
+	return info;
+}
+
+/*
+ * About detectWeb() in this translator ...
+ *
+ * The NDL Online pages do not push state at the completion of page load.
+ * As a result, in search pages, and in item pages accessed by clicking
+ * through from a search list, the URL exposed in the DOM is only a stub,
+ * without the "detail" or "search" elements needed to detect the type of
+ * resource. Also, when clicking through from a search list, the URL stub
+ * does not change, so detectWeb() does not run automatically.
+ *
+ * The state associated with the stub URL is only a tiny skeleton with
+ * no meaningful content. To get the full loaded page, we need to use
+ * monitorDOMChanges on the entire body element, tracking all children.
+ * That works, but it gives rise to an additional problem.
+ *
+ * If this monitorDOMChanges() is applied to item pages accessed via
+ * selectItems(), it forces a rerun of detectWeb() in the middle of the
+ * call, which clobbers the selection so that nothing is downloaded.
+ *
+ * The solution is to attempt to identify single items by their URL
+ * as a fallback, applying monitorDOMChanges() to the page only when
+ * absolutely necessary. When URL detection works, there is no need
+ * to acquire the full DOM, since the URL includes the key needed
+ * to retrieve the raw JSON data that we use for translation.
+ */
+
+function detectWeb(doc, url) {
+	// URL is not properly updated, and doc is initially delivered as a bare skeleton.
+	// To make things work, we need to monitor ALL changes to body.
+	var info = checkPageType(doc, url);
+	if (info.multiple) {
+		return "multiple";
+	} else if (info.single) {
+		return "journalArticle";
+	} else {
+		var body = ZU.xpath(doc, "//body")[0];
+		Z.monitorDOMChanges(body, {childList: true, subtree: true});
+		if (info.single) {
+			return "journalArticle";
+		}
+	}
```

Comment:
I don't think we need to put these lines in the else case, because in the other two cases we return something and will not continue here. Moreover, I think we can delete the second check for `info.single` because when the DOM changes then the function `detectWeb` will be called again.